### PR TITLE
Issue 2476 - Update hzn dev service new/start/stop to support secrets

### PR DESCRIPTION
**Pull request by MaxMcAdam**
Signed-off-by: Max McAdam <max@fredcom.com>

### CODE REVIEW

File: cli/native_deployment/native_deployment.go [hunk 1]
```diff
@@ -311,6 +312,30 @@ func (p *NativeDeploymentConfigPlugin) StartTest(homeDirectory string, userInput
 		cliutils.Fatal(cliutils.CLI_GENERAL_ERROR, msgPrinter.Sprintf("'%v %v' unable to get service dependencies, %v", dev.SERVICE_COMMAND, dev.SERVICE_START_COMMAND, derr))
 	}
 
+
+	// Add binds for the provided service secret files if the secret name is specified in a dependent service
```
Comment (dabooz) [inline cli/native_deployment/native_deployment.go:316]: Move all this logic into a function, probably a stand-alone function in the dev package.

File: cli/native_deployment/native_deployment.go [hunk 2]
```diff
@@ -343,6 +368,17 @@ func (p *NativeDeploymentConfigPlugin) StartTest(homeDirectory string, userInput
 		cliutils.Fatal(cliutils.CLI_GENERAL_ERROR, "'%v %v' %v", dev.SERVICE_COMMAND, dev.SERVICE_START_COMMAND, cerr)
 	}
 
+	// Add binds for the provided service secret files to the top-level service container if the secret name is in the service deployment
```
Comment (dabooz) [inline cli/native_deployment/native_deployment.go:371]: Make this a reusable function and call it from the other new function.

**Review by dabooz — CHANGES_REQUESTED** [review]

File: cli/hzn.go [hunk 3]
```diff
@@ -249,6 +249,7 @@ Environment Variables:
...
 	devServiceNoFSS := devServiceStartTestCmd.Flag("noFSS", msgPrinter.Sprintf("Do not bring up file sync service (FSS) containers. They are brought up by default.")).Short('S').Bool()
 	devServiceStartCmdUserPw := devServiceStartTestCmd.Flag("user-pw", msgPrinter.Sprintf("Horizon Exchange user credentials to query exchange resources. Specify it when you want to automatically fetch the missing dependent services from the Exchange. The default is HZN_EXCHANGE_USER_AUTH environment variable. If you don't prepend it with the user's org, it will automatically be prepended with the value of the HZN_ORG_ID environment variable.")).Short('u').PlaceHolder("USER:PW").String()
+	devServiceStartSecretsFiles := devServiceStartTestCmd.Flag("secret", msgPrinter.Sprintf("Absolute filepath for secrets required by the service or it's dependent services. Filename must match secret name. This flag can be repeated.")).Strings()
```
[the remark this answers is not in the document]
Comment (MaxMcAdam) [reply cli/hzn.go]: Changed

File: cli/dev/service.go [hunk 4]
```diff
@@ -193,10 +194,19 @@ func verifyNewServiceInputs(homeDirectory string, org string, specRef string, ve
 	return dir, nil
 }
 
-func ServiceStartTest(homeDirectory string, userInputFile string, configFiles []string, configType string, noFSS bool, userCreds string) {
+func  mapSecNameToSecPath(secretPaths []string) map[string]string {
```
[the remark this answers is not in the document]
Comment (MaxMcAdam) [reply cli/dev/service.go]: Done

File: cli/dev/service.go [hunk 5]
```diff
@@ -10,6 +10,7 @@ import (
 	"github.com/open-horizon/anax/semanticversion"
 	"os"
 	"runtime"
+	"path/filepath"
```
[the remark this answers is not in the document]
Comment (MaxMcAdam) [reply cli/dev/service.go:12]: Done

File: cli/native_deployment/native_deployment.go [hunk 6]
```diff
@@ -311,6 +312,30 @@ func (p *NativeDeploymentConfigPlugin) StartTest(homeDirectory string, userInput
...
+			fmt.Printf("Failed to marshal deps: %v",err)
+		} else if err = json.Unmarshal(deployBytes, &typedDeploy); err != nil {
+			fmt.Printf("Failed to unmarshal deps: %v",err)
+		} else {
+			for svcName, svcInfo := range typedDeploy.Services {
+				for secName, _ := range svcInfo.Secrets {
+					if svcPath, ok := secretsFiles[secName]; ok {
+						svcInfo.Binds = append(svcInfo.Binds, fmt.Sprintf("%v:%v/%v", svcPath, config.HZN_SECRETS_MOUNT, secName))
+					} else {
+						fmt.Sprintf("Warning: Secret %v for service %v not found.", secName, svcName)
```
[the remark this answers is not in the document]
Comment (MaxMcAdam) [reply cli/native_deployment/native_deployment.go]: Fixed

File: cli/native_deployment/native_deployment.go (hunk 1 above)
Comment (MaxMcAdam) [reply cli/native_deployment/native_deployment.go]: It's in dev/secrets.go now

File: cli/native_deployment/native_deployment.go (hunk 2 above)
Comment (MaxMcAdam) [reply cli/native_deployment/native_deployment.go]: Also in dev/secrets.go now

File: cli/dev/service.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dabooz) [reply cli/dev/service.go:12]: You added a skeletal file for secrets, which is good, but I also wanted the deployment config in the service def to have an empty secret section; `"secrets":{}` is good enough.

File: cli/native_deployment/native_deployment.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dabooz) [reply cli/native_deployment/native_deployment.go]: you missed AddTopLevelServiceSecretBinds()

**Review by dabooz — APPROVED** [review]

**Pull request merged**